### ISSUE TITLE
Manual Upgrade Step to Latest Version (v4.5)

### ISSUE THREAD

**Issue by athomasmoz** [mozilla-mobile/firefox-tv#2794]
### Why/User Benefit/User Problem

On certain FOS version, users will not be auto-upgraded to the latest FFTV version because we've updated permissions to support receive tab. We want to ensure users are upgrading to get new features, enhancements & bug fixes

### Acceptance Criteria (how do I know when I’m done?)

* Users are asked to upgrade to our latest version, if not automatically upgraded
* Users can put off upgrading with a "not now" option, but we're not blocked from forcing update in future

#### Option 1:
Rely on OS update messaging

#### Option 2:
Create our own with design below

Strings
* Heading: Keep Firefox up to date
* Body: Update Firefox for the latest in performance and security. This update lets you send tabs from other Firefox browsers and view them here on the big screen. 
* Button 1: Update
* Button 2: Not now

User flow
https://whimsical.com/Mbrga1N3KwHiJpExhSR23m#2Ux7TurymLxGuX2NAjke

![upgrade-user-flow](https://user-images.githubusercontent.com/51007/64301931-265b1d80-cfd5-11e9-8d78-db6eb314186d.png)

Mocks
https://brampitoyo.github.io/people/bpitoyo/firefox-for-fire-tv/receive-tabs/i03/#artboard0

We already have a header-body-buttons layout used for Turbo Mode onboarding (https://github.com/mozilla-mobile/firefox-tv/pull/2791#issue-314191233). Let’s use the margin/padding/font size provided by that layout, but centre the text vertically.

![Homescreen - Update dialogue@2x](https://user-images.githubusercontent.com/51007/64300996-28bc7800-cfd3-11e9-9dfb-65bfb23677e0.png)

**Comment by brampitoyo** [comment]
@athomasmoz I’ve updated the initial comment with strings, user flow and mockup.

**Comment by dnarcese** [comment]
@athomasmoz Do we want to only show this screen for the users who are **not** being auto-upgraded? So it would be dependent on device

**Comment by athomasmoz** [comment]
I was under the assumption that users who are auto-upgraded would skip the interim version and go right to the newest, since they auto-upgrade to the latest

**Comment by dnarcese** [comment]
> I was under the assumption that users who are auto-upgraded would skip the interim version and go right to the newest, since they auto-upgrade to the latest

@athomasmoz Yes that's true, but there might be some time before the auto-upgrade - I'm not sure how fast it happens

**Comment by dnarcese** [comment]
From https://developer.amazon.com/docs/app-submission/update-published-app.html#app_update_process:
> When customers launch an app that has a manual update pending, Fire TV shows a message that asks if customers want to update the app first before launching it. If customers decline, Fire TV does not prompt customers to update the app on subsequent launches for another seven days.

If we can test this out and it works as we intend, maybe we don't need to implement this ourselves.

**Comment by mcomella** [comment]
One concern I have about releasing two builds in quick succession: if the Amazon store only provides upgrades to users for the newest version, users may not get the opportunity to update to the intermediate version – i.e. the one with our custom upgrade message.

**Comment by athomasmoz** [comment]
Sharing my input since I'm out next week. My preference would be to go with option 1 because it's more simple and I feel there's risk in playing with the upgrade path for users by adding interim versions. Although we still have users on older versions of the app from the last time manual upgrade was needed, this situation is a little different:
* It should only effect users on certain OS versions
* This time, the OS provides a prompt every week asking users to upgrade
* We're not completely sure if our permission changes require manual upgrade

That being said, I think we've come up with a very good alternative option here. If UX & eng decide that approach isn't too risky and makes more sense overall, I'm fine to go that route.

**Comment by brampitoyo** [comment]
@dnarcese @npark-mozilla I agree with @athomasmoz suggestion that we should let Fire OS take care of the update prompt. We don’t need to write or implement our own.

Is there any way for us to trigger a system update prompt that does not contain the “Launch without updating” option? In other words, the only option is to “Update”.

**Comment by npark-mozilla** [comment]
@brampitoyo I don't think we have option to disable "Launch without updating," but if the user has turned on Automatic Update, then the user would never see this dialog and the app will be updated automatically.

**Comment by mcomella** [comment]
> Is there any way for us to trigger a system update prompt that does not contain the “Launch without updating” option? In other words, the only option is to “Update”.

I agree with No-Jun: I doubt it. As some alternatives to creating our own dialog, we could:
- send our own push notifications but they're not super noticeable and we'd have to write code to implement it. tbh, a dialog might be easier
- add a not-super intrusive toast every time the app is opened that says, "Firefox is outdated. Please update" or something. Maybe make it scary by saying it's missing security updates :P

> @brampitoyo I don't think we have option to disable "Launch without updating," but if the user has turned on Automatic Update, then the user would never see this dialog and the app will be updated automatically.

I don't think this is true: as far as I am aware, if new permissions are introduced, all users on Fire OS 5 have to explicitly accept the upgrade, irrespective of the automatic update setting (which makes me wonder if, when we enable automatic updates, we can reproduce this behavior on the LAT).

**Comment by npark-mozilla** [comment]
> I don't think this is true: as far as I am aware, if new permissions are introduced, all users on Fire OS 5 have to explicitly accept the upgrade, irrespective of the automatic update setting (which makes me wonder if, when we enable automatic updates, we can reproduce this behavior on the LAT).

Hmm that's interesting, so automatic updates aren't really automatic? or would the explicit acceptance need to be done only once, and from then on it'll be updated automatically?

**Comment by dnarcese** [comment]
For Fire OS 5, updates are automatic for every update that does not have a new permission. 
For Fire OS 6, all updates are automatic even if there is a new permission. The permission acceptance will happen at runtime.

**Comment by dnarcese** [comment]
We have decided to implement the dialog and release it behind an experiment by the end of this week so that it has time to get to all users before our Send Tab update goes out. 
Testing whether the permissions will require a manual update became larger scope than expected, and so we are going to stop that work and instead spend the time implementing the dialog.

**Comment by Baron-Severin** [comment]
Hey @brampitoyo, I have a few styling/behavior questions on this one.  Note that I've implemented "request update" and "force update" screens, and we can control which is active from our backend.

1) Should the copy be the same between the request and force upgrade screens?
2) How frequently should the request screen be shown? If it would help, we could set multiple options and change these from our backend as well

**Comment by brampitoyo** [comment]
@Baron-Severin 

> Should the copy be the same between the request and force upgrade screens?

Yes. It should be the same. The only difference is the buttons. Request update will have “Not now”. Force update won’t have it.

### Question

Since we will already show a dismiss-able dialogue at the OS-level, does it seem strange to get another dialogue at the app-level that’s _also dismiss-able_?

<img width="1030" alt="Screen Shot 2019-09-12 at 12 14 18 PM" src="https://user-images.githubusercontent.com/51007/64744198-f96bb500-d556-11e9-859d-a2357ef389cd.png">

-----

What if we use this behaviour instead?

<img width="674" alt="Screen Shot 2019-09-12 at 12 23 46 PM" src="https://user-images.githubusercontent.com/51007/64744497-2f5d6900-d558-11e9-976d-084946661b88.png">

Source: https://whimsical.com/Mbrga1N3KwHiJpExhSR23m#2Ux7TurymLtV7ZrfCFqL

* Always show Fire OS dialogue
    * This dialogue can be dismissed
    * IF dismissed, THEN show dialogue again in 7 days
* IF the user hasn’t updated in x weeks, THEN show Firefox dialogue
    * This dialogue cannot be dismissed

-----

> How frequently should the request screen be shown? If it would help, we could set multiple options and change these from our backend as well.

Based on above, I don’t think we should show a “Request update” screen, since we have asked Fire OS to take care of that? Their dialogue is set to show every 7 days when dismissed, and I think that’s sensible.

But on our end, we should show the “Force update” screen after around 4 weeks. And this screen sticks – it can’t be dismissed.
